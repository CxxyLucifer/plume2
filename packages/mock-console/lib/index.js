"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var MockConsole = /** @class */ (function () {
    function MockConsole() {
        this._logs = [];
        this.mock();
    }
    MockConsole.prototype.mock = function () { };
    MockConsole.prototype.mockFn = function () {
        var _this = this;
        return jest.fn(function (log) { return _this._logs.push(log); });
    };
    Object.defineProperty(MockConsole.prototype, "logs", {
        get: function () {
            return this._logs;
        },
        enumerable: true,
        configurable: true
    });
    MockConsole.prototype.reset = function () {
        this._logs = [];
    };
    return MockConsole;
}());
exports.MockConsole = MockConsole;
var MockWarn = /** @class */ (function (_super) {
    __extends(MockWarn, _super);
    function MockWarn() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    MockWarn.prototype.mock = function () {
        console.warn = this.mockFn();
    };
    return MockWarn;
}(MockConsole));
exports.MockWarn = MockWarn;
var MockLog = /** @class */ (function (_super) {
    __extends(MockLog, _super);
    function MockLog() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    MockLog.prototype.mock = function () {
        console.log = this.mockFn();
    };
    return MockLog;
}(MockConsole));
exports.MockLog = MockLog;
