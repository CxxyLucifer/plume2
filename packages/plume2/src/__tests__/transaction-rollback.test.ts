import { MockConsole } from 'mock-console';
import { Action, Actor, IMap, Store } from '../index';

class HelloActor extends Actor {
  defaultState() {
    return { text: 'hello' };
  }

  @Action('change')
  //@ts-ignore
  change(state: IMap, text) {
    throw new Error('change exception');
    // return state.set('text', text);
  }
}

class LoadingActor extends Actor {
  defaultState() {
    return { loading: false };
  }

  @Action('loading:end')
  change(state: IMap) {
    return state.update('loading', loading => !loading);
  }
}

class AppStore extends Store {
  bindActor() {
    return [new LoadingActor(), new HelloActor()];
  }

  rollBack = () => {
    const isRollback = this.transaction(() => {
      this.dispatch('loading:end');
      this.dispatch('change', 'hello iflux2');
    });

    expect(isRollback).toEqual(true);
  };

  customRollBack = () => {
    const currentState = this.state();
    const isRollback = this.transaction(
      () => {
        this.dispatch('loading:end');
        this.dispatch('change', 'hello iflux2');
      },
      () => {
        expect(currentState != this.state()).toEqual(true);
        this.setState(currentState);
      }
    );

    expect(isRollback).toEqual(true);
  };

  change = () => {
    this.transaction(() => {
      try {
        this.dispatch('loading:end');
        this.dispatch('change', 'hello iflux2');
      } catch (err) {
        expect(err).toEqual(new Error('change exception'));
      }
    });
  };
}

describe('test store transaction fail rollback', () => {
  it('test rollback', () => {
    const store = new AppStore();
    const state = store.state();
    store.rollBack();
    expect(store.state()).toEqual(state);
  });

  it('test customRollBack', () => {
    const store = new AppStore({});
    const state = store.state();
    store.customRollBack();
    expect(store.state()).toEqual(state);
  });

  it('test without rollback', () => {
    const mock = new MockConsole();
    const store = new AppStore({ debug: true });
    const state = store.state();
    store.change();
    expect(store.state() == state).toEqual(false);
    expect(mock.logs).toEqual([]);
  });
});
