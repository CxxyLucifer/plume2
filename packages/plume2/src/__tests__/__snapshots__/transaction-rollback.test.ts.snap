// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test store transaction fail rollback test customRollBack 1`] = `
Array [
  "Store init ðŸš€",
  "ðŸ˜­, some exception occur in transaction, store state roll back",
  [Error: change exception],
]
`;

exports[`test store transaction fail rollback test rollback 1`] = `
Array [
  "Store init ðŸš€",
  "ðŸ˜­, some exception occur in transaction, store state roll back",
  [Error: change exception],
]
`;
